ext/nio4r/org/nio4r/ByteBuffer.java

```diff
@@ -0,0 +1,283 @@
+        }
+        return bytes;
+    }
+
+    @JRubyMethod(name = "write_to")
+    public IRubyObject writeTo(ThreadContext context, IRubyObject f) {
+        try {
+            File file = (File) JavaUtil.unwrapJavaObject(f);
+            if (!isTheSameFile(file, false)) {
+                currentWritePath = file.getAbsolutePath();
+                if (currentWriteFileChannel != null) currentWriteFileChannel.close();
+                if (fileOutputStream != null) fileOutputStream.close();
+
+                fileOutputStream = new FileOutputStream(file, true);
+                currentWriteFileChannel = fileOutputStream.getChannel();
+            }
+            currentWriteFileChannel.write(byteBuffer);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("File Write Operation Error:  " + e.getLocalizedMessage());
+        }
+        return this;
+    }
+
+    @JRubyMethod(name = "read_from")
+    public IRubyObject readFrom(ThreadContext context, IRubyObject f) {
+        try {
+            File file = (File) JavaUtil.unwrapJavaObject(f);
+            if (!isTheSameFile(file, true)) {
+                inChannel.close();
+                currentReadChannel.close();
+                currentReadPath = file.getAbsolutePath();
+                currentReadChannel = new FileInputStream(file);
+                inChannel = currentReadChannel.getChannel();
+            }
+            inChannel.read(byteBuffer);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("File Read Operation Error:  " + e.getLocalizedMessage());
+        }
+        return this;
+    }
+
+    private boolean isTheSameFile(File f, boolean read) {
+        if (read)
+            return (currentReadPath == f.getAbsolutePath());
+        return currentWritePath == f.getAbsolutePath();
+    }
+
+    @JRubyMethod(name = "remaining")
+    public IRubyObject remainingPositions(ThreadContext context) {
+        Ruby ruby = context.getRuntime();
+        int count = byteBuffer.remaining();
+        return RubyNumeric.int2fix(ruby, count);
+    }
+
+    @JRubyMethod(name = "remaining?")
+    public IRubyObject hasRemaining(ThreadContext context) {
+        if (byteBuffer.hasRemaining())
+            return context.getRuntime().getTrue();
+
+        return context.getRuntime().getFalse();
+    }
+
+    @JRubyMethod(name = "offset?")
+    public IRubyObject getOffset(ThreadContext context) {
+        int offset = byteBuffer.arrayOffset();
+        return JavaUtil.convertJavaToRuby(context.getRuntime(), offset);
+    }
+
+    /**
+     * Check whether the two ByteBuffers are the same.
+     *
+     * @param context
+     * @param ob      : The RubyObject which needs to be check
+     * @return
+     */
+    @JRubyMethod(name = "equals?")
+    public IRubyObject equals(ThreadContext context, IRubyObject ob) {
+        boolean equal = this.byteBuffer.equals(((ByteBuffer) JavaUtil.convertRubyToJava(ob)).getBuffer());
```

Comment:
should `equals?` really fail if not a wrapped `ByteBuffer` object is passed (possibly with a `ClassCastException`) ?